### PR TITLE
Update serializers technote and ChapelIO

### PR DESCRIPTION
**Pull request by benharsh**
This PR updates the serializers technote to include the full Serializer and Deserializer API, as well as information concerning the relevant interfaces (e.g. ``writeSerializable``).

### CODE REVIEW

File: modules/standard/ChapelIO.chpl [hunk 1]
```diff
@@ -38,11 +38,171 @@ of a Hello World program:
...
+
+  var p = new Point(4, 2);
+
+  // Prints '(4, 2)' in the default serialization format
+  stdout.writeln(p);
+
+  // Prints '[4, 2]' in the JSON serialization format
+  var jsonWriter = stdout.withSerializer(jsonSerializer);
+  jsonWriter.writeln(p);
+
```
Comment (jeremiah-corrado) [inline modules/standard/ChapelIO.chpl:151]: I might add a sentence here along the lines of: "an analogous ``deserialize`` method or deserializing initializer could also be defined on `Point` such that calls to `read(Point)` would attempt to read it as a tuple"

**Review by jeremiah-corrado — APPROVED** [review]
Looks great!

I have some suggestions about wording/clarity:

File: doc/rst/technotes/ioSerializers.rst [hunk 2]
```diff
@@ -283,6 +535,17 @@ method on records and classes. If a suitable ``deserialize`` method is not
...
+For classes, the ``deserializeValue`` method has the freedom to potentially
+free the given class and/or reassign it, depending on the needs of the
+Deserializer.
+
+The arguments ``val`` or ``readType`` are defined to be a "primitive" type or a
+type that implements at least one of the following interfaces:
+
+- ``readDeserializable``
+- ``initDeserializable``
+- ``serializable``
```
[the remark this answers is not in the document]
Comment (benharsh) [reply doc/rst/technotes/ioSerializers.rst:547]: I believe this is correct as "serializable". ``serializable`` is the catch-all interface that satisfies all three of the other interfaces. I'll add a note along those lines.

[the remark this answers is not in the document]
Comment (benharsh) [reply doc/rst/technotes/ioSerializers.rst:547]: We could add a ``deserializable`` at some point, but that hasn't been discussed yet.

[the remark this answers is not in the document]
Comment (jeremiah-corrado) [reply doc/rst/technotes/ioSerializers.rst:547]: Oh that's right, I forgot about that

**Review by jeremiah-corrado — APPROVED** [review]

**Pull request merged**